samza-core/src/main/java/org/apache/samza/operators/impl/WatermarkStates.java

```diff
@@ -71,6 +71,8 @@ synchronized void update(long timestamp, String taskName) {
       if (timestamps.size() == expectedTotal) {
         Optional<Long> min = timestamps.values().stream().min(Long::compare);
         watermarkTime = min.orElse(timestamp);
+      } else if (taskName == null) {
```

Comment:
Please clarify the execution path here + comments.

Reply:
Fixed! Thanks for pointing this out.